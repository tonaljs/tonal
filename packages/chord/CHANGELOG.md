# @tonaljs/chord

## 4.6.10

### Patch Changes

- Bug fixing
- Updated dependencies
  - @tonaljs/chord-detect@4.6.10
  - @tonaljs/chord-type@4.6.10
  - @tonaljs/collection@4.6.10
  - @tonaljs/core@4.6.10
  - @tonaljs/pcset@4.6.10
  - @tonaljs/scale-type@4.6.10
