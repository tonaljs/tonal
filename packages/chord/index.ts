import { detect } from "@tonaljs/chord-detect";
import {
  all as chordTypes,
  ChordType,
  get as getChordType,
} from "@tonaljs/chord-type";
import { tonicIntervalsTransposer } from "@tonaljs/core";

import {
  deprecate,
  distance,
  note,
  NoteName,
  tokenizeNote,
  transpose as transposeNote,
} from "@tonaljs/core";

import { isSubsetOf, isSupersetOf } from "@tonaljs/pcset";

import { all as scaleTypes } from "@tonaljs/scale-type";
export { detect } from "@tonaljs/chord-detect";

type ChordName = string;
type ChordNameTokens = [string, string]; // [TONIC, SCALE TYPE]

export interface Chord extends ChordType {
  tonic: string | null;
  type: string;
  root: string;
  rootDegree: number;
  symbol: string;
  notes: NoteName[];
}

const NoChord: Chord = {
  empty: true,
  name: "",
  symbol: "",
  root: "",
  rootDegree: 0,
  type: "",
  tonic: null,
  setNum: NaN,
  quality: "Unknown",
  chroma: "",
  normalized: "",
  aliases: [],
  notes: [],
  intervals: [],
};

// 6, 64, 7, 9, 11 and 13 are consider part of the chord
// (see https://github.com/danigb/tonal/issues/55)
const NUM_TYPES = /^(6|64|7|9|11|13)$/;
/**
 * Tokenize a chord name. It returns an array with the tonic and chord type
 * If not tonic is found, all the name is considered the chord name.
 *
 * This function does NOT check if the chord type exists or not. It only tries
 * to split the tonic and chord type.
 *
 * @function
 * @param {string} name - the chord name
 * @return {Array} an array with [tonic, type]
 * @example
 * tokenize("Cmaj7") // => [ "C", "maj7" ]
 * tokenize("C7") // => [ "C", "7" ]
 * tokenize("mMaj7") // => [ null, "mMaj7" ]
 * tokenize("Cnonsense") // => [ null, "nonsense" ]
 */
export function tokenize(name: string): ChordNameTokens {
  const [letter, acc, oct, type] = tokenizeNote(name);
  if (letter === "") {
    return ["", name];
  }
  // aug is augmented (see https://github.com/danigb/tonal/issues/55)
  if (letter === "A" && type === "ug") {
    return ["", "aug"];
  }
  // see: https://github.com/tonaljs/tonal/issues/70
  if (!type && (oct === "4" || oct === "5")) {
    return [letter + acc, oct];
  }

  if (NUM_TYPES.test(oct)) {
    return [letter + acc, oct + type];
  } else {
    return [letter + acc + oct, type];
  }
}

/**
 * Get a Chord from a chord name.
 */
export function get(src: ChordName | ChordNameTokens): Chord {
  if (src === "") {
    return NoChord;
  }
  if (Array.isArray(src) && src.length === 2) {
    return getChord(src[1], src[0]);
  } else {
    const [tonic, type] = tokenize(src);
    const chord = getChord(type, tonic);
    return chord.empty ? getChord(src) : chord;
  }
}

/**
 * Get chord properties
 *
 * @param typeName - the chord type name
 * @param [tonic] - Optional tonic
 * @param [root]  - Optional root (requires a tonic)
 */
export function getChord(
  typeName: string,
  optionalTonic?: string,
  optionalRoot?: string
): Chord {
  const type = getChordType(typeName);
  const tonic = note(optionalTonic || "");
  const root = note(optionalRoot || "");

  if (
    type.empty ||
    (optionalTonic && tonic.empty) ||
    (optionalRoot && root.empty)
  ) {
    return NoChord;
  }

  const rootInterval = distance(tonic.pc, root.pc);
  const rootDegree = type.intervals.indexOf(rootInterval) + 1;
  if (!root.empty && !rootDegree) {
    return NoChord;
  }

  const intervals = Array.from(type.intervals);

  for (let i = 1; i < rootDegree; i++) {
    const num = intervals[0][0];
    const quality = intervals[0][1];
    const newNum = parseInt(num, 10) + 7;
    intervals.push(`${newNum}${quality}`);
    intervals.shift();
  }

  const notes = tonic.empty
    ? []
    : intervals.map((i) => transposeNote(tonic, i));

  typeName = type.aliases.indexOf(typeName) !== -1 ? typeName : type.aliases[0];
  const symbol = `${tonic.empty ? "" : tonic.pc}${typeName}${
    root.empty || rootDegree <= 1 ? "" : "/" + root.pc
  }`;
  const name = `${optionalTonic ? tonic.pc + " " : ""}${type.name}${
    rootDegree > 1 && optionalRoot ? " over " + root.pc : ""
  }`;
  return {
    ...type,
    name,
    symbol,
    type: type.name,
    root: root.name,
    intervals,
    rootDegree,
    tonic: tonic.name,
    notes,
  };
}

export const chord = deprecate("Chord.chord", "Chord.get", get);

/**
 * Transpose a chord name
 *
 * @param {string} chordName - the chord name
 * @return {string} the transposed chord
 *
 * @example
 * transpose('Dm7', 'P4') // => 'Gm7
 */
export function transpose(chordName: string, interval: string): string {
  const [tonic, type] = tokenize(chordName);
  if (!tonic) {
    return chordName;
  }
  return transposeNote(tonic, interval) + type;
}

/**
 * Get all scales where the given chord fits
 *
 * @example
 * chordScales('C7b9')
 * // => ["phrygian dominant", "flamenco", "spanish heptatonic", "half-whole diminished", "chromatic"]
 */
export function chordScales(name: string): string[] {
  const s = get(name);
  const isChordIncluded = isSupersetOf(s.chroma);
  return scaleTypes()
    .filter((scale) => isChordIncluded(scale.chroma))
    .map((scale) => scale.name);
}
/**
 * Get all chords names that are a superset of the given one
 * (has the same notes and at least one more)
 *
 * @function
 * @example
 * extended("CMaj7")
 * // => [ 'Cmaj#4', 'Cmaj7#9#11', 'Cmaj9', 'CM7add13', 'Cmaj13', 'Cmaj9#11', 'CM13#11', 'CM7b9' ]
 */
export function extended(chordName: string): string[] {
  const s = get(chordName);
  const isSuperset = isSupersetOf(s.chroma);
  return chordTypes()
    .filter((chord) => isSuperset(chord.chroma))
    .map((chord) => s.tonic + chord.aliases[0]);
}

/**
 * Find all chords names that are a subset of the given one
 * (has less notes but all from the given chord)
 *
 * @example
 */
export function reduced(chordName: string): string[] {
  const s = get(chordName);
  const isSubset = isSubsetOf(s.chroma);
  return chordTypes()
    .filter((chord) => isSubset(chord.chroma))
    .map((chord) => s.tonic + chord.aliases[0]);
}

/**
 * Returns a function to get a note name from the scale degree.
 *
 * @example
 * [1, 2, 3, 4].map(Chord.degrees("C")) => ["C", "E", "G", "C"]
 * [1, 2, 3, 4].map(Chord.degrees("C4")) => ["C4", "E4", "G4", "C5"]
 */
export function degrees(chordName: string | ChordNameTokens) {
  const { intervals, tonic } = get(chordName);
  const transpose = tonicIntervalsTransposer(intervals, tonic);
  return (degree: number) =>
    degree ? transpose(degree > 0 ? degree - 1 : degree) : "";
}

/**
 * Sames as `degree` but with 0-based index
 */
export function steps(chordName: string | ChordNameTokens) {
  const { intervals, tonic } = get(chordName);
  return tonicIntervalsTransposer(intervals, tonic);
}

export default {
  getChord,
  get,
  detect,
  chordScales,
  extended,
  reduced,
  tokenize,
  transpose,
  degrees,
  steps,

  // deprecate
  chord,
};
