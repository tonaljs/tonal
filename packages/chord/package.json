{
  "name": "@tonaljs/chord",
  "version": "5.0.0",
  "description": "Musical chords and its relations",
  "keywords": [
    "chord",
    "chords",
    "music",
    "theory",
    "@tonaljs"
  ],
  "main": "dist/index.js",
  "module": "dist/index.mjs",
  "files": [
    "dist"
  ],
  "types": "dist/index.d.ts",
  "dependencies": {
    "@tonaljs/chord-detect": "^4.8.2",
    "@tonaljs/chord-type": "^5.0.0",
    "@tonaljs/collection": "^4.8.0",
    "@tonaljs/core": "^4.10.0",
    "@tonaljs/pcset": "^4.8.2",
    "@tonaljs/scale-type": "^4.8.2"
  },
  "author": "danigb@gmail.com",
  "license": "MIT",
  "publishConfig": {
    "access": "public"
  },
  "scripts": {
    "build": "tsup index.ts --sourcemap --dts --format esm,cjs"
  }
}
