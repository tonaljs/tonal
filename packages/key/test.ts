// tslint:disable-next-line: no-implicit-dependencies
import { get as scale } from "@tonaljs/scale";
// tslint:disable-next-line: no-implicit-dependencies
import { get as chord } from "@tonaljs/chord";
import Key from "./index";

describe("@tonal/key", () => {
  test("fromAlter", () => {
    expect(Key.majorTonicFromKeySignature("###")).toEqual("A");
    expect(Key.majorTonicFromKeySignature(3)).toEqual("A");
    expect(Key.majorTonicFromKeySignature("b")).toEqual("F");
    expect(Key.majorTonicFromKeySignature("bb")).toEqual("Bb");
    expect(Key.majorTonicFromKeySignature("other")).toEqual(null);
  });
  test("major keySignature", () => {
    const tonics = "C D E F G A B".split(" ");
    expect(
      tonics.map((tonic) => Key.majorKey(tonic).keySignature).join(" ")
    ).toEqual(" ## #### b # ### #####");
  });

  test("minor keySignature", () => {
    const tonics = "C D E F G A B".split(" ");
    expect(
      tonics.map((tonic) => Key.minorKey(tonic).keySignature).join(" ")
    ).toEqual("bbb b # bbbb bb  ##");
  });

  describe("scale names", () => {
    test("natural scales", () => {
      const chordScales = Key.minorKey("C").natural.chordScales;
      expect(chordScales.map(scale).map((scale) => scale.name)).toEqual([
        "C minor",
        "D locrian",
        "Eb major",
        "F dorian",
        "G phrygian",
        "Ab lydian",
        "Bb mixolydian",
      ]);
    });

    test("harmonic scales", () => {
      const chordScales = Key.minorKey("C").harmonic.chordScales;
      expect(chordScales.map(scale).map((scale) => scale.name)).toEqual([
        "C harmonic minor",
        "D locrian 6",
        "Eb major augmented",
        "F lydian diminished",
        "G phrygian dominant",
        "Ab lydian #9",
        "B ultralocrian",
      ]);
    });

    test("melodic scales", () => {
      const chordScales = Key.minorKey("C").melodic.chordScales;
      expect(chordScales.map(scale).map((scale) => scale.name)).toEqual([
        "C melodic minor",
        "D dorian b2",
        "Eb lydian augmented",
        "F lydian dominant",
        "G mixolydian b6",
        "A locrian #2",
        "B altered",
      ]);
    });
  });

  test("secondary dominants", () => {
    expect(Key.majorKey("C").secondaryDominants).toEqual([
      "",
      "A7",
      "B7",
      "C7",
      "D7",
      "E7",
      "",
    ]);
  });

  test("octaves are discarded", () => {
    expect(Key.majorKey("b4").scale.join(" ")).toEqual("B C# D# E F# G# A#");
    expect(Key.majorKey("g4").chords.join(" ")).toEqual(
      "Gmaj7 Am7 Bm7 Cmaj7 D7 Em7 F#m7b5"
    );
    expect(Key.minorKey("C4").melodic.scale.join(" ")).toEqual(
      "C D Eb F G A B"
    );
    expect(Key.minorKey("C4").melodic.chords.join(" ")).toEqual(
      "Cm6 Dm7 Eb+maj7 F7 G7 Am7b5 Bm7b5"
    );
  });

  test("valid chord names", () => {
    const major = Key.majorKey("C");
    const minor = Key.minorKey("C");

    [
      major.chords,
      major.secondaryDominants,
      major.secondaryDominantsMinorRelative,
      major.substituteDominants,
      major.substituteDominantsMinorRelative,
      minor.natural.chords,
      minor.harmonic.chords,
      minor.melodic.chords,
    ].forEach((chords) => {
      chords.forEach((name) => {
        if (name !== "") {
          if (chord(name).name === "") throw Error(`Invalid chord: ${name}`);
          expect(chord(name).name).not.toBe("");
        }
      });
    });
  });

  test("majorKey", () => {
    expect(Key.majorKey("C")).toMatchInlineSnapshot(`
      {
        "alteration": 0,
        "chordScales": [
          "C major",
          "D dorian",
          "E phrygian",
          "F lydian",
          "G mixolydian",
          "A minor",
          "B locrian",
        ],
        "chords": [
          "Cmaj7",
          "Dm7",
          "Em7",
          "Fmaj7",
          "G7",
          "Am7",
          "Bm7b5",
        ],
        "chordsHarmonicFunction": [
          "T",
          "SD",
          "T",
          "SD",
          "D",
          "T",
          "D",
        ],
        "grades": [
          "I",
          "II",
          "III",
          "IV",
          "V",
          "VI",
          "VII",
        ],
        "intervals": [
          "1P",
          "2M",
          "3M",
          "4P",
          "5P",
          "6M",
          "7M",
        ],
        "keySignature": "",
        "minorRelative": "A",
        "scale": [
          "C",
          "D",
          "E",
          "F",
          "G",
          "A",
          "B",
        ],
        "secondaryDominants": [
          "",
          "A7",
          "B7",
          "C7",
          "D7",
          "E7",
          "",
        ],
        "secondaryDominantsMinorRelative": [
          "",
          "Em7b5",
          "F#m7",
          "Gm7",
          "Am7",
          "Bm7b5",
          "",
        ],
        "substituteDominants": [
          "",
          "Eb7",
          "F7",
          "Gb7",
          "Ab7",
          "Bb7",
          "",
        ],
        "substituteDominantsMinorRelative": [
          "",
          "Em7",
          "Cm7",
          "Dbm7",
          "Am7",
          "Fm7",
          "",
        ],
        "tonic": "C",
        "triads": [
          "C",
          "Dm",
          "Em",
          "F",
          "G",
          "Am",
          "Bdim",
        ],
        "type": "major",
      }
    `);
  });

  test("empty major key ", () => {
    expect(Key.majorKey("")).toMatchObject({
      type: "major",
      tonic: "",
    });
    expect(Object.keys(Key.majorKey("C")).sort()).toEqual(
      Object.keys(Key.majorKey("")).sort()
    );
  });

  test("minorKey", () => {
    expect(Key.minorKey("C")).toMatchInlineSnapshot(`
      {
        "alteration": -3,
        "harmonic": {
          "chordScales": [
            "C harmonic minor",
            "D locrian 6",
            "Eb major augmented",
            "F lydian diminished",
            "G phrygian dominant",
            "Ab lydian #9",
            "B ultralocrian",
          ],
          "chords": [
            "CmMaj7",
            "Dm7b5",
            "Eb+maj7",
            "Fm7",
            "G7",
            "Abmaj7",
            "Bo7",
          ],
          "chordsHarmonicFunction": [
            "T",
            "SD",
            "T",
            "SD",
            "D",
            "SD",
            "D",
          ],
          "grades": [
            "I",
            "II",
            "bIII",
            "IV",
            "V",
            "bVI",
            "VII",
          ],
          "intervals": [
            "1P",
            "2M",
            "3m",
            "4P",
            "5P",
            "6m",
            "7M",
          ],
          "scale": [
            "C",
            "D",
            "Eb",
            "F",
            "G",
            "Ab",
            "B",
          ],
          "tonic": "C",
          "triads": [
            "Cm",
            "Ddim",
            "Ebaug",
            "Fm",
            "G",
            "Ab",
            "Bdim",
          ],
        },
        "keySignature": "bbb",
        "melodic": {
          "chordScales": [
            "C melodic minor",
            "D dorian b2",
            "Eb lydian augmented",
            "F lydian dominant",
            "G mixolydian b6",
            "A locrian #2",
            "B altered",
          ],
          "chords": [
            "Cm6",
            "Dm7",
            "Eb+maj7",
            "F7",
            "G7",
            "Am7b5",
            "Bm7b5",
          ],
          "chordsHarmonicFunction": [
            "T",
            "SD",
            "T",
            "SD",
            "D",
            "",
            "",
          ],
          "grades": [
            "I",
            "II",
            "bIII",
            "IV",
            "V",
            "VI",
            "VII",
          ],
          "intervals": [
            "1P",
            "2M",
            "3m",
            "4P",
            "5P",
            "6M",
            "7M",
          ],
          "scale": [
            "C",
            "D",
            "Eb",
            "F",
            "G",
            "A",
            "B",
          ],
          "tonic": "C",
          "triads": [
            "Cm",
            "Dm",
            "Ebaug",
            "F",
            "G",
            "Adim",
            "Bdim",
          ],
        },
        "natural": {
          "chordScales": [
            "C minor",
            "D locrian",
            "Eb major",
            "F dorian",
            "G phrygian",
            "Ab lydian",
            "Bb mixolydian",
          ],
          "chords": [
            "Cm7",
            "Dm7b5",
            "Ebmaj7",
            "Fm7",
            "Gm7",
            "Abmaj7",
            "Bb7",
          ],
          "chordsHarmonicFunction": [
            "T",
            "SD",
            "T",
            "SD",
            "D",
            "SD",
            "SD",
          ],
          "grades": [
            "I",
            "II",
            "bIII",
            "IV",
            "V",
            "bVI",
            "bVII",
          ],
          "intervals": [
            "1P",
            "2M",
            "3m",
            "4P",
            "5P",
            "6m",
            "7m",
          ],
          "scale": [
            "C",
            "D",
            "Eb",
            "F",
            "G",
            "Ab",
            "Bb",
          ],
          "tonic": "C",
          "triads": [
            "Cm",
            "Ddim",
            "Eb",
            "Fm",
            "Gm",
            "Ab",
            "Bb",
          ],
        },
        "relativeMajor": "Eb",
        "tonic": "C",
        "type": "minor",
      }
    `);
  });

  test("empty minor key ", () => {
    expect(Key.minorKey("nothing")).toMatchObject({
      type: "minor",
      tonic: "",
    });
    expect(Object.keys(Key.minorKey("C")).sort()).toEqual(
      Object.keys(Key.minorKey("nothing")).sort()
    );
  });
});
