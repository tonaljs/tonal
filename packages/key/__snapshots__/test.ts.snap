// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`@tonal/key > C minor chords 1`] = `
[
  {
    "name": "Cm7",
    "roles": [
      "T",
    ],
  },
  {
    "name": "Dm7b5",
    "roles": [
      "SD",
    ],
  },
  {
    "name": "Ebmaj7",
    "roles": [
      "T",
    ],
  },
  {
    "name": "Fm7",
    "roles": [
      "SD",
    ],
  },
  {
    "name": "Gm7",
    "roles": [
      "D",
      "ii/IV",
    ],
  },
  {
    "name": "Abmaj7",
    "roles": [
      "SD",
    ],
  },
  {
    "name": "Bb7",
    "roles": [
      "SD",
    ],
  },
  {
    "name": "G7",
    "roles": [
      "V/I",
      "D",
    ],
  },
  {
    "name": "C7",
    "roles": [
      "V/IV",
    ],
  },
  {
    "name": "D7",
    "roles": [
      "V/V",
    ],
  },
  {
    "name": "Eb7",
    "roles": [
      "V/bVI",
      "subV/II",
    ],
  },
  {
    "name": "F7",
    "roles": [
      "V/bVII",
      "SD",
    ],
  },
  {
    "name": "Dm7",
    "roles": [
      "ii/I",
      "SD",
    ],
  },
  {
    "name": "Am7",
    "roles": [
      "ii/V",
    ],
  },
  {
    "name": "Bbm7b5",
    "roles": [
      "ii/bVI",
    ],
  },
  {
    "name": "Cm7b5",
    "roles": [
      "ii/bVII",
    ],
  },
  {
    "name": "Db7",
    "roles": [
      "subV/I",
    ],
  },
  {
    "name": "Gb7",
    "roles": [
      "subV/IV",
    ],
  },
  {
    "name": "Ab7",
    "roles": [
      "subV/V",
    ],
  },
  {
    "name": "Bbb7",
    "roles": [
      "subV/bVI",
    ],
  },
  {
    "name": "Cb7",
    "roles": [
      "subV/bVII",
    ],
  },
  {
    "name": "Abm7",
    "roles": [
      "subii/I",
    ],
  },
  {
    "name": "Dbm7",
    "roles": [
      "subii/IV",
    ],
  },
  {
    "name": "Ebm7",
    "roles": [
      "subii/V",
    ],
  },
  {
    "name": "Fbm7b5",
    "roles": [
      "subii/bVI",
    ],
  },
  {
    "name": "Gbm7b5",
    "roles": [
      "subii/bVII",
    ],
  },
  {
    "name": "CmMaj7",
    "roles": [
      "T",
    ],
  },
  {
    "name": "Eb+maj7",
    "roles": [
      "T",
    ],
  },
  {
    "name": "Bo7",
    "roles": [
      "D",
    ],
  },
  {
    "name": "Am7b5",
    "roles": [
      "ii/V",
    ],
  },
  {
    "name": "Ebm7b5",
    "roles": [
      "subii/V",
    ],
  },
  {
    "name": "Cm6",
    "roles": [
      "T",
    ],
  },
  {
    "name": "Bm7b5",
    "roles": [],
  },
  {
    "name": "A7",
    "roles": [
      "V/II",
    ],
  },
  {
    "name": "Em7",
    "roles": [
      "ii/II",
    ],
  },
  {
    "name": "Gm7b5",
    "roles": [
      "ii/IV",
    ],
  },
  {
    "name": "Bbm7",
    "roles": [
      "subii/II",
    ],
  },
  {
    "name": "Dbm7b5",
    "roles": [
      "subii/IV",
    ],
  },
]
`;

exports[`@tonal/key C major chords 1`] = `
[
  {
    "name": "Cmaj7",
    "roles": [
      "T",
    ],
  },
  {
    "name": "Dm7",
    "roles": [
      "SD",
    ],
  },
  {
    "name": "Em7",
    "roles": [
      "T",
      "ii/II",
    ],
  },
  {
    "name": "Fmaj7",
    "roles": [
      "SD",
    ],
  },
  {
    "name": "G7",
    "roles": [
      "D",
    ],
  },
  {
    "name": "Am7",
    "roles": [
      "T",
    ],
  },
  {
    "name": "Bm7b5",
    "roles": [
      "D",
    ],
  },
  {
    "name": "A7",
    "roles": [
      "V/II",
    ],
  },
  {
    "name": "B7",
    "roles": [
      "V/III",
    ],
  },
  {
    "name": "C7",
    "roles": [
      "V/IV",
    ],
  },
  {
    "name": "D7",
    "roles": [
      "V/V",
    ],
  },
  {
    "name": "E7",
    "roles": [
      "V/VI",
    ],
  },
  {
    "name": "F#m7",
    "roles": [
      "ii/III",
    ],
  },
  {
    "name": "Gm7b5",
    "roles": [
      "ii/IV",
    ],
  },
  {
    "name": "Am7b5",
    "roles": [
      "ii/V",
    ],
  },
  {
    "name": "Bm7",
    "roles": [
      "ii/VI",
    ],
  },
  {
    "name": "Eb7",
    "roles": [
      "subV/II",
    ],
  },
  {
    "name": "F7",
    "roles": [
      "subV/III",
    ],
  },
  {
    "name": "Gb7",
    "roles": [
      "subV/IV",
    ],
  },
  {
    "name": "Ab7",
    "roles": [
      "subV/V",
    ],
  },
  {
    "name": "Bb7",
    "roles": [
      "subV/VI",
    ],
  },
  {
    "name": "Bbm7",
    "roles": [
      "subii/II",
    ],
  },
  {
    "name": "Cm7",
    "roles": [
      "subii/III",
    ],
  },
  {
    "name": "Dbm7b5",
    "roles": [
      "subii/IV",
    ],
  },
  {
    "name": "Ebm7b5",
    "roles": [
      "subii/V",
    ],
  },
  {
    "name": "Fm7",
    "roles": [
      "subii/VI",
    ],
  },
]
`;

exports[`@tonal/key C minor chords 1`] = `
[
  {
    "name": "Cm7",
    "roles": [
      "T",
    ],
  },
  {
    "name": "Dm7b5",
    "roles": [
      "SD",
    ],
  },
  {
    "name": "Ebmaj7",
    "roles": [
      "T",
    ],
  },
  {
    "name": "Fm7",
    "roles": [
      "SD",
    ],
  },
  {
    "name": "Gm7",
    "roles": [
      "D",
      "ii/IV",
    ],
  },
  {
    "name": "Abmaj7",
    "roles": [
      "SD",
    ],
  },
  {
    "name": "Bb7",
    "roles": [
      "SD",
    ],
  },
  {
    "name": "G7",
    "roles": [
      "V/I",
      "D",
    ],
  },
  {
    "name": "C7",
    "roles": [
      "V/IV",
    ],
  },
  {
    "name": "D7",
    "roles": [
      "V/V",
    ],
  },
  {
    "name": "Eb7",
    "roles": [
      "V/bVI",
      "subV/II",
    ],
  },
  {
    "name": "F7",
    "roles": [
      "V/bVII",
      "SD",
    ],
  },
  {
    "name": "Dm7",
    "roles": [
      "ii/I",
      "SD",
    ],
  },
  {
    "name": "Am7",
    "roles": [
      "ii/V",
    ],
  },
  {
    "name": "Bbm7b5",
    "roles": [
      "ii/bVI",
    ],
  },
  {
    "name": "Cm7b5",
    "roles": [
      "ii/bVII",
    ],
  },
  {
    "name": "Db7",
    "roles": [
      "subV/I",
    ],
  },
  {
    "name": "Gb7",
    "roles": [
      "subV/IV",
    ],
  },
  {
    "name": "Ab7",
    "roles": [
      "subV/V",
    ],
  },
  {
    "name": "Bbb7",
    "roles": [
      "subV/bVI",
    ],
  },
  {
    "name": "Cb7",
    "roles": [
      "subV/bVII",
    ],
  },
  {
    "name": "Abm7",
    "roles": [
      "subii/I",
    ],
  },
  {
    "name": "Dbm7",
    "roles": [
      "subii/IV",
    ],
  },
  {
    "name": "Ebm7",
    "roles": [
      "subii/V",
    ],
  },
  {
    "name": "Fbm7b5",
    "roles": [
      "subii/bVI",
    ],
  },
  {
    "name": "Gbm7b5",
    "roles": [
      "subii/bVII",
    ],
  },
  {
    "name": "CmMaj7",
    "roles": [
      "T",
    ],
  },
  {
    "name": "Eb+maj7",
    "roles": [
      "T",
    ],
  },
  {
    "name": "Bo7",
    "roles": [
      "D",
    ],
  },
  {
    "name": "Am7b5",
    "roles": [
      "ii/V",
    ],
  },
  {
    "name": "Ebm7b5",
    "roles": [
      "subii/V",
    ],
  },
  {
    "name": "Cm6",
    "roles": [
      "T",
    ],
  },
  {
    "name": "Bm7b5",
    "roles": [],
  },
  {
    "name": "A7",
    "roles": [
      "V/II",
    ],
  },
  {
    "name": "Em7",
    "roles": [
      "ii/II",
    ],
  },
  {
    "name": "Gm7b5",
    "roles": [
      "ii/IV",
    ],
  },
  {
    "name": "Bbm7",
    "roles": [
      "subii/II",
    ],
  },
  {
    "name": "Dbm7b5",
    "roles": [
      "subii/IV",
    ],
  },
]
`;

exports[`@tonal/key majorKeys 1`] = `
{
  "alteration": 3,
  "chordScales": [
    "A major",
    "B dorian",
    "C# phrygian",
    "D lydian",
    "E mixolydian",
    "F# minor",
    "G# locrian",
  ],
  "chords": [
    "Amaj7",
    "Bm7",
    "C#m7",
    "Dmaj7",
    "E7",
    "F#m7",
    "G#m7b5",
  ],
  "chordsHarmonicFunction": [
    "T",
    "SD",
    "T",
    "SD",
    "D",
    "T",
    "D",
  ],
  "grades": [
    "I",
    "II",
    "III",
    "IV",
    "V",
    "VI",
    "VII",
  ],
  "intervals": [
    "1P",
    "2M",
    "3M",
    "4P",
    "5P",
    "6M",
    "7M",
  ],
  "keySignature": "###",
  "minorRelative": "F#",
  "scale": [
    "A",
    "B",
    "C#",
    "D",
    "E",
    "F#",
    "G#",
  ],
  "secondaryDominantSupertonics": [
    "",
    "C#m7",
    "D#m7",
    "Em7b5",
    "F#m7b5",
    "G#m7",
    "",
  ],
  "secondaryDominants": [
    "",
    "F#7",
    "G#7",
    "A7",
    "B7",
    "C#7",
    "",
  ],
  "secondaryDominantsMinorRelative": [
    "",
    "C#m7",
    "D#m7",
    "Em7b5",
    "F#m7b5",
    "G#m7",
    "",
  ],
  "substituteDominantSupertonics": [
    "",
    "Gm7",
    "Am7",
    "Bbm7b5",
    "Cm7b5",
    "Dm7",
    "",
  ],
  "substituteDominants": [
    "",
    "C7",
    "D7",
    "Eb7",
    "F7",
    "G7",
    "",
  ],
  "substituteDominantsMinorRelative": [
    "",
    "Gm7",
    "Am7",
    "Bbm7b5",
    "Cm7b5",
    "Dm7",
    "",
  ],
  "tonic": "A",
  "triads": [
    "A",
    "Bm",
    "C#m",
    "D",
    "E",
    "F#m",
    "G#dim",
  ],
  "type": "major",
}
`;

exports[`@tonal/key majorKeys 2`] = `
{
  "alteration": -2,
  "chordScales": [
    "Bb major",
    "C dorian",
    "D phrygian",
    "Eb lydian",
    "F mixolydian",
    "G minor",
    "A locrian",
  ],
  "chords": [
    "Bbmaj7",
    "Cm7",
    "Dm7",
    "Ebmaj7",
    "F7",
    "Gm7",
    "Am7b5",
  ],
  "chordsHarmonicFunction": [
    "T",
    "SD",
    "T",
    "SD",
    "D",
    "T",
    "D",
  ],
  "grades": [
    "I",
    "II",
    "III",
    "IV",
    "V",
    "VI",
    "VII",
  ],
  "intervals": [
    "1P",
    "2M",
    "3M",
    "4P",
    "5P",
    "6M",
    "7M",
  ],
  "keySignature": "bb",
  "minorRelative": "G",
  "scale": [
    "Bb",
    "C",
    "D",
    "Eb",
    "F",
    "G",
    "A",
  ],
  "secondaryDominantSupertonics": [
    "",
    "Dm7",
    "Em7",
    "Fm7b5",
    "Gm7b5",
    "Am7",
    "",
  ],
  "secondaryDominants": [
    "",
    "G7",
    "A7",
    "Bb7",
    "C7",
    "D7",
    "",
  ],
  "secondaryDominantsMinorRelative": [
    "",
    "Dm7",
    "Em7",
    "Fm7b5",
    "Gm7b5",
    "Am7",
    "",
  ],
  "substituteDominantSupertonics": [
    "",
    "Abm7",
    "Bbm7",
    "Cbm7b5",
    "Dbm7b5",
    "Ebm7",
    "",
  ],
  "substituteDominants": [
    "",
    "Db7",
    "Eb7",
    "Fb7",
    "Gb7",
    "Ab7",
    "",
  ],
  "substituteDominantsMinorRelative": [
    "",
    "Abm7",
    "Bbm7",
    "Cbm7b5",
    "Dbm7b5",
    "Ebm7",
    "",
  ],
  "tonic": "Bb",
  "triads": [
    "Bb",
    "Cm",
    "Dm",
    "Eb",
    "F",
    "Gm",
    "Adim",
  ],
  "type": "major",
}
`;

exports[`@tonal/key majorKeys 3`] = `
{
  "alteration": 4,
  "chordScales": [
    "E major",
    "F# dorian",
    "G# phrygian",
    "A lydian",
    "B mixolydian",
    "C# minor",
    "D# locrian",
  ],
  "chords": [
    "Emaj7",
    "F#m7",
    "G#m7",
    "Amaj7",
    "B7",
    "C#m7",
    "D#m7b5",
  ],
  "chordsHarmonicFunction": [
    "T",
    "SD",
    "T",
    "SD",
    "D",
    "T",
    "D",
  ],
  "grades": [
    "I",
    "II",
    "III",
    "IV",
    "V",
    "VI",
    "VII",
  ],
  "intervals": [
    "1P",
    "2M",
    "3M",
    "4P",
    "5P",
    "6M",
    "7M",
  ],
  "keySignature": "####",
  "minorRelative": "C#",
  "scale": [
    "E",
    "F#",
    "G#",
    "A",
    "B",
    "C#",
    "D#",
  ],
  "secondaryDominantSupertonics": [
    "",
    "G#m7",
    "A#m7",
    "Bm7b5",
    "C#m7b5",
    "D#m7",
    "",
  ],
  "secondaryDominants": [
    "",
    "C#7",
    "D#7",
    "E7",
    "F#7",
    "G#7",
    "",
  ],
  "secondaryDominantsMinorRelative": [
    "",
    "G#m7",
    "A#m7",
    "Bm7b5",
    "C#m7b5",
    "D#m7",
    "",
  ],
  "substituteDominantSupertonics": [
    "",
    "Dm7",
    "Em7",
    "Fm7b5",
    "Gm7b5",
    "Am7",
    "",
  ],
  "substituteDominants": [
    "",
    "G7",
    "A7",
    "Bb7",
    "C7",
    "D7",
    "",
  ],
  "substituteDominantsMinorRelative": [
    "",
    "Dm7",
    "Em7",
    "Fm7b5",
    "Gm7b5",
    "Am7",
    "",
  ],
  "tonic": "E",
  "triads": [
    "E",
    "F#m",
    "G#m",
    "A",
    "B",
    "C#m",
    "D#dim",
  ],
  "type": "major",
}
`;
